### PR TITLE
feat: added configurable prefixes to api

### PR DESCRIPTION
**Pull request by franciscog-kainos**

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import uk.gov.companieshouse.config.CompanyNumberConfiguration;
+
+@Component
+public class CompanyNumberValidator implements ConstraintValidator<ValidCompanyNumber, String> {
+
+    private final Pattern prefixListRegex = Pattern.compile("(?i)^([A-Z][A-Z]?)\\b(,[A-Z][A-Z]?)*$");
+
+    private final Pattern companyNumberRegex;
+
+    public CompanyNumberValidator(CompanyNumberConfiguration prefixConfig) {
```
Comment (n-r-b) [inline src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java:21]: what's the reason for having this constructor? Can we not add this logic to the `isValid` method?

File: src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import javax.validation.ConstraintValidatorContext;
+
+import org.springframework.stereotype.Component;
+
+import uk.gov.companieshouse.config.CompanyNumberConfiguration;
+
+@Component
+public class CompanyNumberValidator implements ConstraintValidator<ValidCompanyNumber, String> {
+
+    private final Pattern prefixListRegex = Pattern.compile("(?i)^([A-Z][A-Z]?)\\b(,[A-Z][A-Z]?)*$");
```
Comment (n-r-b) [inline src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java:17]: can we made this static?

Comment (franciscog-kainos) [reply src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java:17]: What's the benefit of that?

File: src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java (hunk 1 above)
Comment (franciscog-kainos) [reply src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java:21]: Constructor dependency injection. The program will fail earlier at runtime if the config is incorrect rather than when the method is called

File: src/main/java/uk/gov/companieshouse/config/CompanyNumberConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package uk.gov.companieshouse.config;
+
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class CompanyNumberConfiguration {
+    @Value("${companyNumber.prefixes}")
```
Comment (n-r-b) [inline src/main/java/uk/gov/companieshouse/config/CompanyNumberConfiguration.java:8]: At the moment we only have one property, we could just inject this into the validator?

File: src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+        this.companyNumberRegex = Pattern.compile("(?i)^("
+                + String.join("|", List.of(singleCharacterPrefixRegex, doubleCharacterPrefixRegex, onlyNumbersRegex))
+                + ")$");
+
+    }
+
+    @Override
+    public boolean isValid(String value, ConstraintValidatorContext context) {
+        return this.companyNumberRegex.matcher(value).matches();
```
Comment (n-r-b) [inline src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java:49]: I would advise breaking this up so it's more readable, example below (variable and method names could be improved :P):

```
    @Override
    public boolean isValid(String value, ConstraintValidatorContext context) {

        boolean prefixIsValid = arePrefixesValid();
        boolean companyNumberIsValid = isCompanyNumberValid(value);

        return prefixIsValid && companyNumberIsValid;
    }
```

File: src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    private final Pattern prefixListRegex = Pattern.compile("(?i)^([A-Z][A-Z]?)\\b(,[A-Z][A-Z]?)*$");
+
+    private final Pattern companyNumberRegex;
+
+    public CompanyNumberValidator(CompanyNumberConfiguration prefixConfig) {
+
+        String prefixString = prefixConfig.getPrefixes();
+
+        if (!this.prefixListRegex.matcher(prefixString).matches()) {
+            throw new IllegalArgumentException(
```
Comment (n-r-b) [inline src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java:26]: Do we really want to fail here? We could return false (for invalid) and add logging to explain the prefix config is invalid

File: src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java (hunk 2 above)
Comment (n-r-b) [reply src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java:16]: this being a static constant allows you to access it without creating an instance of the class and it's common to all objects of the class

File: src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+                    "Prefix list formatting error. Make sure list is comma separated e.g. NI,SI,R. Current: "
+                            + prefixString);
+        }
+
+        this.companyNumberRegex = Pattern.compile("(?i)^(" + String.join("|", generatePrefixList(prefixString)) + ")$");
+
+    }
+
+    @Override
+    public boolean isValid(String value, ConstraintValidatorContext context) {
```
Comment (n-r-b) [inline src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java:33]: what if the value null?

File: src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java (hunk 2 above)
Comment (n-r-b) [reply src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java:16]: just noticed this is only used in one place... we could hide only the regex behind a constant

File: src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+import org.springframework.stereotype.Component;
+
+@Component
+public class CompanyNumberValidator implements ConstraintValidator<ValidCompanyNumber, String> {
+
+    private final Pattern prefixListRegex = Pattern.compile("(?i)^([A-Z][A-Z]?)\\b(,[A-Z][A-Z]?)*$");
+
+    private final Pattern companyNumberRegex;
+
+    public CompanyNumberValidator(@Value("${companyNumber.prefixes}") String prefixString) {
```
Comment (n-r-b) [inline src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java:18]: what happens if this config does not exist?

Comment (franciscog-kainos) [reply src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java:18]: Standard behaviour. It will throw an exception

File: src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    @Override
+    public boolean isValid(String value, ConstraintValidatorContext context) {
+        return this.companyNumberRegex.matcher(value).matches();
+    }
+
+    private List<String> generatePrefixList(String prefixString) {
+        List<String> prefixesArray = Arrays.asList(prefixString.split(","));
+
+        String singleCharacterPrefixRegex = "("
+                .concat(createPrefixSubstring(prefixesArray, 1).concat(")").concat("[0-9]{1,7}"));
```
Comment (n-r-b) [inline src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java:46]: it would be nice to see constants for the raw regex so we know exactly what they're used for

Comment (franciscog-kainos) [reply src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java:46]: What do you mean?

File: src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+        String doubleCharacterPrefixRegex = "("
+                .concat(createPrefixSubstring(prefixesArray, 2).concat(")").concat("[0-9]{1,6}"));
+
+        String onlyNumbersRegex = "[0-9]{1,8}";
+
+        return List.of(singleCharacterPrefixRegex, doubleCharacterPrefixRegex, onlyNumbersRegex);
+    }
+
+    private String safeSubstring(String value, int startIndex) {
```
Comment (n-r-b) [inline src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java:56]: looks like the startIndex is always 1 here

File: src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java (hunk 7 above)
Comment (n-r-b) [reply src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java:19]: instead of creating a constructor, why not override the `initialize` method on the ConstraintValidator?

Comment (franciscog-kainos) [reply src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java:19]: What's the benefit?

File: src/test/java/uk/gov/companieshouse/model/validation/CompanyNumberValidatorTest.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
+package uk.gov.companieshouse.model.validation;
+
+import static org.junit.Assert.assertTrue;
```
Comment (n-r-b) [inline src/test/java/uk/gov/companieshouse/model/validation/CompanyNumberValidatorTest.java:4]: please use the jupiter assertion here

File: src/test/java/uk/gov/companieshouse/model/validation/CompanyNumberValidatorTest.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+@ExtendWith(SpringExtension.class)
+public class CompanyNumberValidatorTest {
+
+    @Test
+    void shouldThrowAnExceptionWhenInputListIsInWrongFormat() {
+
+        List<String> invalidLists = List.of(",", "A,[.", "A.BC.D");
+
+        invalidLists.forEach(prefixList -> {
```
Comment (n-r-b) [inline src/test/java/uk/gov/companieshouse/model/validation/CompanyNumberValidatorTest.java:26]: this can be simplified: 
`invalidLists.forEach(prefixList -> assertThrows(IllegalArgumentException.class, () -> new CompanyNumberValidator(prefixList), prefixList));`

File: src/test/java/uk/gov/companieshouse/model/validation/CompanyNumberValidatorTest.java [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+            assertThrows(IllegalArgumentException.class, () -> new CompanyNumberValidator(prefixList), prefixList);
+        });
+
+    }
+
+    @Test
+    void shouldCreateFactorySuccessfulyWhenValidListsProvided() {
+        List<String> validLists = List.of("NI", "AB,CD,EF", "A,B,C,D,F", "AB,cd,e,F");
+        validLists.forEach(prefixList -> {
```
Comment (n-r-b) [inline src/test/java/uk/gov/companieshouse/model/validation/CompanyNumberValidatorTest.java:35]: can be simplified too, same as above

File: src/test/java/uk/gov/companieshouse/model/validation/CompanyNumberValidatorTest.java [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+    @Test
+    void shouldReturnFalseWhenPrefixIsNotAllowed() {
+        String prefixes = "SC,NI,OC,SO,R,AP";
+
+        CompanyNumberValidator companyNumberValidator = new CompanyNumberValidator(prefixes);
+
+        List<String> invalidCompanyNumbers = List.of("TY000001", "PTABC0123", "XC999999", "T0000001");
+
+        boolean isAllInvalid = invalidCompanyNumbers.stream().map(value -> companyNumberValidator.isValid(value, null))
```
Comment (n-r-b) [inline src/test/java/uk/gov/companieshouse/model/validation/CompanyNumberValidatorTest.java:48]: instead of using `allMatch` when passing false, why not use `noneMatch`:

`boolean isAllInvalid = invalidCompanyNumbers.stream().noneMatch(value -> companyNumberValidator.isValid(value, null));`

File: src/test/java/uk/gov/companieshouse/model/validation/CompanyNumberValidatorTest.java [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+    void shouldReturnTrueWhenPrefixesAreAllowed() {
+        String prefixes = "SC,NI,OC,SO,R,AP";
+
+        CompanyNumberValidator companyNumberValidator = new CompanyNumberValidator(prefixes);
+
+        List<String> validUpperCaseCompanyNumbers = List.of("NI000000", "SC123123", "OC123123", "SO123123", "R0000000",
+                "R123", "123", "AP123456");
+
+        List<String> validLowerCaseCompanyNumbers = validUpperCaseCompanyNumbers.stream()
+                .map(companyNumber -> companyNumber.toLowerCase()).collect(Collectors.toList());
```
Comment (n-r-b) [inline src/test/java/uk/gov/companieshouse/model/validation/CompanyNumberValidatorTest.java:65]: `companyNumber -> companyNumber.toLowerCase()` can be replaced with method reference, i.e. `String::toLowerCase`

File: src/test/java/uk/gov/companieshouse/model/validation/CompanyNumberValidatorTest.java [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+
+        boolean isAllValid = Stream.of(validUpperCaseCompanyNumbers, validLowerCaseCompanyNumbers)
+                .flatMap(x -> x.stream()).map(value -> companyNumberValidator.isValid(value, null))
```
Comment (n-r-b) [inline src/test/java/uk/gov/companieshouse/model/validation/CompanyNumberValidatorTest.java:68]: `x -> x.stream()` can also be `Collection::stream`

Comment (n-r-b) [reply src/test/java/uk/gov/companieshouse/model/validation/CompanyNumberValidatorTest.java:68]: instead of using both `map` and `allMatch`, you can merge the two:
`.allMatch(value -> companyNumberValidator.isValid(value, null));`

File: src/test/java/uk/gov/companieshouse/model/validation/CompanyNumberValidatorTest.java [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+import java.util.stream.Stream;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+
+import uk.gov.companieshouse.model.validator.CompanyNumberValidator;
+
+@ExtendWith(SpringExtension.class)
+public class CompanyNumberValidatorTest {
```
Comment (n-r-b) [inline src/test/java/uk/gov/companieshouse/model/validation/CompanyNumberValidatorTest.java:19]: we should also have a test to cover empty/null prefixes

File: src/test/java/uk/gov/companieshouse/model/validation/CompanyNumberValidatorTest.java [hunk 17]
```diff
@@ -0,0 +1,74 @@
+package uk.gov.companieshouse.model.validation;
```
Comment (n-r-b) [inline src/test/java/uk/gov/companieshouse/model/validation/CompanyNumberValidatorTest.java:1]: please rename package to match one in main

File: src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java (hunk 7 above)
Comment (n-r-b) [reply src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java:19]: well it's there for us to use exactly for this purpose, why define something new?

Comment (franciscog-kainos) [reply src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java:19]: it's there for when you want to create an anonymous class where you won't have access to a constructor right?

Comment (n-r-b) [reply src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java:19]: Screenshot 2020-06-24 at 15 47 25

Comment (n-r-b) [reply src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java:19]: I know your constructor is doing exactly the same, but since we are implementing this class and this method exists (guaranteed to be called), then why not use it for validating the config?

File: src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+@Component
+public class CompanyNumberValidator implements ConstraintValidator<ValidCompanyNumber, String> {
+
+    private final Pattern companyNumberRegex;
+
+    public CompanyNumberValidator(@Value("${companyNumber.prefixes}") String prefixString) {
+
+        Pattern prefixListRegex = Pattern.compile("(?i)^([A-Z][A-Z]?)\\b(,[A-Z][A-Z]?)*$");
+
+        if (!prefixListRegex.matcher(prefixString).matches()) {
```
Comment (n-r-b) [inline src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java:23]: if `prefixString` is null, the regex matcher will throw a null pointer exception, we should ensure this doesn't happen

Comment (franciscog-kainos) [reply src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java:23]: image

File: src/main/java/uk/gov/companieshouse/model/PenaltyIdentifier.java [hunk 19]
```diff
@@ -2,13 +2,15 @@
...
 import javax.validation.constraints.NotBlank;
 import javax.validation.constraints.Pattern;
+
+import uk.gov.companieshouse.model.validator.ValidCompanyNumber;
+
 import java.util.Objects;
 
 public class PenaltyIdentifier {
 
     @NotBlank(message = "companyNumber must not be blank")
```
Comment (n-r-b) [inline src/main/java/uk/gov/companieshouse/model/PenaltyIdentifier.java:12]: is this still needed now we have a custom validator?
